### PR TITLE
Merge 3.3 to main

### PR DESCRIPTION
**Pull request by hpidcock**
Forward ports:
- #16380
- #16390
- #16365
- #16374
- #16405
- #16406
- #16408
- #16409
- #16403
- #16411
- #16412
- #16410

Conflicts:
- apiserver/facades/client/client/client.go
- 🚩 **apiserver/facades/client/controller/controller.go**
- apiserver/facades/client/controller/controller_test.go
- apiserver/facades/controller/caasapplicationprovisioner/provisioner.go
- apiserver/facades/controller/caasmodelconfigmanager/register.go
- apiserver/facades/controller/caasmodeloperator/operator.go
- apiserver/facades/controller/caasoperatorprovisioner/provisioner.go
- apiserver/facades/controller/caasunitprovisioner/provisioner.go
- caas/kubernetes/provider/bootstrap_test.go
- cmd/containeragent/initialize/package_test.go
- controller/config.go
- controller/config_test.go
- internal/docker/auth.go
- worker/caasmodelconfigmanager/worker.go
- worker/caasmodelconfigmanager/worker_test.go

Changes:
- api/controller/caasmodelconfigmanager/client.go
- 🚩 **apiserver/facades/controller/caasmodelconfigmanager/facade.go**
- apiserver/facades/schema.json
- caas/kubernetes/provider/bootstrap.go
- cloudconfig/podcfg/image.go
- cmd/juju/commands/bootstrap.go
- worker/caasmodelconfigmanager/mocks/facade_mock.go

### CODE REVIEW

File: apiserver/facades/client/controller/controller.go [hunk 1]
```diff
@@ -745,6 +746,50 @@ func (c *ControllerAPI) ConfigSet(ctx context.Context, args params.ControllerCon
...
 		return errors.Trace(err)
 	}
+
+	currentCfg, err := c.state.ControllerConfig()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// TODO(dqlite): probably move this garbage logic to domain/controllerconfig/service
+	if newValue, ok := args.Config[corecontroller.CAASImageRepo]; ok {
```
Comment (hpidcock) [inline apiserver/facades/client/controller/controller.go:756]: Review me!

File: apiserver/facades/controller/caasmodelconfigmanager/facade.go [hunk 2]
```diff
@@ -6,17 +6,44 @@ package caasmodelconfigmanager
...
 	return f.controllerConfigAPI.ControllerConfig(ctx)
 }
+
+func (f *Facade) WatchControllerConfig() (params.StringsWatchResult, error) {
+	result := params.StringsWatchResult{}
+	w, err := f.controllerConfigService.Watch()
+	if err != nil {
+		return result, errors.Trace(err)
+	}
+	result.StringsWatcherId, result.Changes, err = internal.EnsureRegisterWatcher[[]string](f.watcherRegistry, w)
```
Comment (hpidcock) [inline apiserver/facades/controller/caasmodelconfigmanager/facade.go:46]: Review me! Am I doing the services right? Maybe?

File: caas/kubernetes/provider/export_test.go [hunk 3]
```diff
@@ -96,7 +96,10 @@ func NewcontrollerStackForTest(
 	pcfg *podcfg.ControllerPodConfig,
 ) (ControllerStackerForTest, error) {
 	cs, err := newcontrollerStack(ctx, stackName, storageClass, broker, pcfg)
-	return cs.(*controllerStack), err
```
Comment (SimonRichardson) [inline caas/kubernetes/provider/export_test.go:99]: Luckily this was just in testing.

**Review by SimonRichardson — APPROVED** [review]
That took some time sorry. I was bringing it back in again after I purged it from my memory.

**Pull request merged**